### PR TITLE
Fixes #20553 Tree flattening should be separate from Fusing

### LINKED ISSUE #20553: Tree flattening should be separate from Fusing
During fusing all GraphStages with AsyncBoundary attribute will create separate groups. These groups are bridged by BoundarySubscriber and BoundaryPublisher. However different materializers may not want to inline the nested computations (GraphStageLogic) done in GraphAssembly.materialize within these groups.

### PR DESCRIPTION
**Pull request by kkasravi**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/Fusing.scala [hunk 1]
```diff
@@ -34,6 +34,12 @@ object Fusing {
   def aggressive[S <: Shape, M](g: Graph[S, M]): FusedGraph[S, M] = Impl.aggressive(g)
 
   /**
+   * Return the StructuralInfo for this Graph without any fusing
+   */
+  def structuralInfo[S <: Shape, M](g: Graph[S, M], attributes: Attributes): StructuralInfoModule =
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/Fusing.scala]: this should probably have an `INTERNAL API` marker in the docs, since `StructuralInfoModule` is in the `impl` package

**Review by patriknw — COMMENTED** [review]
looking good, let's see what @drewhk  says when he is back next week

File: akka-stream/src/main/scala/akka/stream/Fusing.scala (hunk 1 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/Fusing.scala]: Thanks @patriknw - I added the necessary filter in MiMa.scala for the internal type StructuralInfo which is now StructuralInfoModule. Hopefully this build will now pass.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Fusing.scala [hunk 2]
```diff
@@ -443,13 +470,18 @@ private[stream] object Fusing {
...
+      StructuralInfoModule(
+        Set.empty ++ modules.asScala,
+        shape,
         immutable.Map.empty ++ downstreams.asScala,
+        immutable.Map.empty ++ upstreams.asScala,
         immutable.Map.empty ++ inOwners.asScala,
         immutable.Map.empty ++ outOwners.asScala,
-        Set.empty ++ modules.asScala)
+        matValues,
+        matValues.head._2,
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Fusing.scala]: Can `matValues` be guaranteed to be non-empty here?

**Comment by drewhk** [comment]
LGTM, except the issue what @viktorklang found.

**Comment by kkasravi** [comment]
@drewhk @viktorklang I'll fix today and push a new PR. Thanks!

**Comment by kkasravi** [comment]
@drewhk, @viktorklang 

Hi - 

I did the following
```
    case object BogusModule extends AtomicModule {
      override def shape = ClosedShape
      override def replaceShape(s: Shape) =
        if (s != shape) throw new UnsupportedOperationException("cannot replace shape")
        else this

      override def compose(that: Module): Module = compose(that, scaladsl.Keep.left)

      override def compose[A, B, C](that: Module, f: (A, B) => C): Module = {
        if (f eq scaladsl.Keep.right) {
          that
        } else if (f eq scaladsl.Keep.left) {
          val mat =
          if (IgnorableMatValComp(that)) {
            Ignore
          } else {
            Transform(_ => NotUsed, that.materializedValueComputation)
          }
          CompositeModule(
            if (that.isSealed) Set(that) else that.subModules,
            that.shape,
            that.downstreams,
            that.upstreams,
            mat,
            if (this.isSealed) Attributes.none else attributes)
        } else {
          throw new UnsupportedOperationException(
            "It is invalid to combine materialized value with BogusModule " +
            "except with Keep.left or Keep.right")
        }
      }

      override def withAttributes(attributes: Attributes): Module =
        throw new UnsupportedOperationException("BogusModule cannot carry attributes")

      override def attributes = Attributes.none
      override def carbonCopy: Module = this
      override def isRunnable: Boolean = true
      override def isAtomic: Boolean = true
      override def materializedValueComputation: MaterializedValueNode = Ignore
    }

    RunnableGraph(BogusModule).run()
```

The matValues in BuildStructuralInfo.toInfo had a single entry  (BogusModule -> Copy of BogusModule). From my analysis - it's not possible to not have a matValues without an entry  (matValues.head can never be null).

However a scala.MatchError was thrown due to BogusModule not matching any type within ActorMaterializerImpl.materializeAtomic. This is probably ok IMO since BogusModule isn't extending Source, Sink, GraphStage, etc. Stack trace is below:

```
Exception in thread "main" scala.MatchError: BogusModule (of class org.apache.gearpump.akkastream.example.Test9$BogusModule$2$)
	at akka.stream.impl.ActorMaterializerImpl$$anon$2.materializeAtomic(ActorMaterializerImpl.scala:165)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:954)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:950)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:950)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:957)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:950)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:950)
	at akka.stream.impl.MaterializerSession.materialize(StreamLayout.scala:917)
	at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:240)
	at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:138)
	at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:350)
	at org.apache.gearpump.akkastream.example.Test9$.main(Test9.scala:107)
	at org.apache.gearpump.util.AkkaApp$$anonfun$main$1.apply(AkkaApp.scala:42)
	at org.apache.gearpump.util.AkkaApp$$anonfun$main$1.apply(AkkaApp.scala:42)
	at scala.util.Try$.apply(Try.scala:192)
	at org.apache.gearpump.util.AkkaApp$class.main(AkkaApp.scala:41)
	at org.apache.gearpump.akkastream.example.Test9$.main(Test9.scala:36)
	at org.apache.gearpump.akkastream.example.Test9.main(Test9.scala)
```

**Comment by drewhk** [comment]
@kkasravi can you please add a test though that materializes an empty RunnableGraph? Just in case. Otherwise LGTM

**Comment by kkasravi** [comment]
@drewhk thanks - will add the the empty RunnableGraph today and repush.

File: akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala [hunk 3]
```diff
@@ -85,6 +129,11 @@ class FusingSpec extends StreamSpec {
       verify(fused, modules = 2, downstreams = 6)
     }
 
+    "fuse a Module with no subModules" in {
+      val structuralInfoModule = Fusing.structuralInfo(RunnableGraph(NoSubModulesModule), Attributes.none)
+      structuralInfoModule.matValues.size > 0
+    }
+
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala]: Hm, I was actually thinking about trying this:

```scala
val g = GraphDSL.create { b =>
  ClosedShape
}
Fusing.aggressive(g)
```

**Review by drewhk — CHANGES_REQUESTED** [review]
Test an actual empty graph instead of a synthetic module. Then this can be merged.

File: akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala (hunk 3 above)
Comment (kkasravi) [inline akka-stream-tests/src/test/scala/akka/stream/FusingSpec.scala]: Thanks @drewhk - sorry I missed this. I'll update the PR.

**Review by drewhk — APPROVED** [review]
